File: flutter_custom_tabs_android/android/src/main/java/com/github/droibit/flutter/plugins/customtabs/NativeAppLauncher.java
package com.github.droibit.flutter.plugins.customtabs;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.annotation.RestrictTo;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * ref. <a href="https://developer.chrome.com/docs/android/custom-tabs/howto-custom-tab-native-apps/">Let native applications handle the content</a>
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class NativeAppLauncher {
    static boolean launch(@NonNull Context context, @NonNull Uri uri) {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R ?
                launchNativeApi30(context, uri) :
                launchNativeBeforeApi30(context, uri);
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private static boolean launchNativeApi30(@NonNull Context context, @NonNull Uri uri) {
        final Intent nativeAppIntent = new Intent(Intent.ACTION_VIEW, uri)
                .addCategory(Intent.CATEGORY_BROWSABLE)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER);
        try {
            context.startActivity(nativeAppIntent);
            return true;
        } catch (ActivityNotFoundException ignored) {
            return false;
        }
    }

    private static boolean launchNativeBeforeApi30(@NonNull Context context, @NonNull Uri uri) {
        final PackageManager pm = context.getPackageManager();

        // Get all Apps that resolve a generic url
        final Intent browserActivityIntent = new Intent()
                .setAction(Intent.ACTION_VIEW)
                .addCategory(Intent.CATEGORY_BROWSABLE)
                .setData(Uri.fromParts(uri.getScheme(), "", null));
        final Set<String> genericResolvedList =
                extractPackageNames(queryIntentActivities(pm, browserActivityIntent));

        // Get all apps that resolve the specific Url
        final Intent specializedActivityIntent = new Intent(Intent.ACTION_VIEW, uri)
                .addCategory(Intent.CATEGORY_BROWSABLE);
        final Set<String> resolvedSpecializedList =
                extractPackageNames(queryIntentActivities(pm, specializedActivityIntent));

        // Keep only the Urls that resolve the specific, but not the generic urls.
        resolvedSpecializedList.removeAll(genericResolvedList);
        // If the list is empty, no native app handlers were found.
        if (resolvedSpecializedList.isEmpty()) {
            return false;
        }

        // We found native handlers. Launch the Intent.
        specializedActivityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(specializedActivityIntent);
        return true;
    }

    @SuppressWarnings("deprecation")
    private static @NonNull List<ResolveInfo> queryIntentActivities(
            @NonNull PackageManager pm,
            @NonNull Intent intent
    ) {
        final int flags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
                ? PackageManager.MATCH_ALL : PackageManager.MATCH_DEFAULT_ONLY;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return pm.queryIntentActivities(
                    intent,
                    PackageManager.ResolveInfoFlags.of(flags)
            );
        } else {
            return pm.queryIntentActivities(intent, flags);
        }
    }

    private static @NonNull Set<String> extractPackageNames(@NonNull List<ResolveInfo> resolveInfo) {
        final Set<String> packageNames = new HashSet<>(resolveInfo.size());
        for (ResolveInfo info : resolveInfo) {
            packageNames.add(info.activityInfo.packageName);
        }
        return packageNames;
    }

}

File: flutter_custom_tabs_android/android/src/main/kotlin/com/github/droibit/flutter/plugins/customtabs/core/NativeAppLauncher.kt
package com.github.droibit.flutter.plugins.customtabs.core

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.net.Uri
import android.os.Build
import androidx.annotation.RequiresApi

/**
 * ref. [Let native applications handle the content](https://developer.chrome.com/docs/android/custom-tabs/howto-custom-tab-native-apps/)
 */
class NativeAppLauncher {
    fun launch(context: Context, uri: Uri): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            launchNativeApi30(context, uri)
        } else {
            launchNativeBeforeApi30(context, uri)
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private fun launchNativeApi30(context: Context, uri: Uri): Boolean {
        val nativeAppIntent = Intent(Intent.ACTION_VIEW, uri)
            .addCategory(Intent.CATEGORY_BROWSABLE)
            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER)
        try {
            context.startActivity(nativeAppIntent)
            return true
        } catch (ignored: ActivityNotFoundException) {
            return false
        }
    }

    private fun launchNativeBeforeApi30(context: Context, uri: Uri): Boolean {
        val pm = context.packageManager

        // Get all Apps that resolve a generic url
        val browserActivityIntent = Intent()
            .setAction(Intent.ACTION_VIEW)
            .addCategory(Intent.CATEGORY_BROWSABLE)
            .setData(Uri.fromParts(uri.scheme, "", null))
        val genericResolvedList: Set<String> = extractPackageNames(
            queryIntentActivities(pm, browserActivityIntent)
        )

        // Get all apps that resolve the specific Url
        val specializedActivityIntent = Intent(Intent.ACTION_VIEW, uri)
            .addCategory(Intent.CATEGORY_BROWSABLE)
        val resolvedSpecializedList = buildSet {
            addAll(
                extractPackageNames(queryIntentActivities(pm, specializedActivityIntent))
            )
            // Keep only the Urls that resolve the specific, but not the generic urls.
            removeAll(genericResolvedList)
        }
        // If the list is empty, no native app handlers were found.
        if (resolvedSpecializedList.isEmpty()) {
            return false
        }

        // We found native handlers. Launch the Intent.
        specializedActivityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(specializedActivityIntent)
        return true
    }

    @Suppress("deprecation")
    private fun queryIntentActivities(pm: PackageManager, intent: Intent): List<ResolveInfo> {
        val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PackageManager.MATCH_ALL
        } else {
            PackageManager.MATCH_DEFAULT_ONLY
        }

        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            pm.queryIntentActivities(
                intent,
                PackageManager.ResolveInfoFlags.of(flags.toLong())
            )
        } else {
            pm.queryIntentActivities(intent, flags)
        }
    }

    private fun extractPackageNames(resolveInfo: List<ResolveInfo>): Set<String> {
        return buildSet(resolveInfo.size) {
            for (info in resolveInfo) {
                add(info.activityInfo.packageName)
            }
        }
    }
}
